server/pulp/server/db/model/__init__.py

```diff
@@ -307,3 +312,132 @@ def post_save(cls, sender, document, **kwargs):
+    storage_path = StringField(db_field='_storage_path')
+
+    # For backward compatibility
+    _ns = StringField(required=True)
+    unit_type_id = StringField(db_field='_content_type_id', required=True)
+    _source_location = None
+    _relative_path = None
+
+    meta = {
+        'abstract': True,
+    }
+
+    @classmethod
+    def post_init_signal(cls, sender, document):
+        """
+        The signal that is triggered before a unit is initialized
+
+        This is used to validate that the unit_key_fields attribute is set properly
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+
+        if not hasattr(document, 'unit_key_fields'):
+            class_name = type(document).__name__
+            raise exceptions.PulpCodedException(error_codes.PLP0035, class_name=class_name)
+
+    @classmethod
+    def pre_save_signal(cls, sender, document, **kwargs):
+        """
+        The signal that is triggered before a unit is saved
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+        if not document.id:
+            document.id = str(uuid.uuid4())
+        document.last_updated = dateutils.now_utc_timestamp()
+
+    @classmethod
+    def post_save_signal(cls, sender, document, **kwargs):
+        """
+        The signal that is triggered after a unit is saved
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+        # If content was set on this unit, move the content into place
+        if document._source_location and document._relative_path:
+            server_storage_dir = config.config.get('server', 'storage_dir')
+            platform_storage_dir = os.path.join(server_storage_dir, 'content',
+                                                document.unit_type_id)
+            target_location = os.path.join(platform_storage_dir, document._relative_path)
+            # Make sure the directory for the target exists
+            if os.path.isdir(document._source_location):
+                shutil.copytree(document._source_location, target_location)
+            else:
+                check_dir = os.path.dirname(target_location)
+                misc.mkdir(check_dir)
+                # Copy the content into place
+                shutil.copy(document._source_location, target_location)
+
+    def set_content(self, source_location, relative_path):
+        """
+        Store the source of the content for the unit and the relative path
+        where it should be stored within the plugin content directory
+
+        :param source_location: The absolute path to the content in the plugin working directory
+        :type source_location: str
```

Comment:
I'm not sure if this is a good suggestion, but maybe `basestr` instead for these since unicode should work for these.